### DISCUSSION TITLE
Element Integral Variable user object

### DISCUSSION THREAD

**Discussion by gabburgio** [idaholab/moose#25055] [Q&A General] [answered]
Hello everybody

I wanted to ask a question about a detail I don't understand in the implementation of the Element Integral / Element Integral Variable user object.

The `virtual Real getValue();` method in the Element Integral user object is meant to provide the value of the postprocessor for external use, if I understand correctly.
However, if I try to do something like:
`_uo(getUserObject<ElementIntegralVariableUserObject>("user_object")),`
to retrieve the user object, and then call `uo.getValue()`, for instance in some method of the class I'm working on, I get an error of the type: 
`passing 'const ElementIntegralVariableUserObject' as 'this' argument discards qualifiers [-fpermissive]`

This , to my understanding, is due to the fact that `getValue()` is not a constant method, therefore it cannot be called on a constant reference like the one returned by `getUserObject`.

My question is then, why is `getValue()` not a constant method?
I see for instance that the public return methods of LayeredBase are constant:

```
 virtual Real integralValue(Point p) const;

  /**
   * Get the value for a given layer
   * @param layer The layer index
   * @return The value for the given layer
   */
  virtual Real getLayerValue(unsigned int layer) const;

  /**
   * Helper function to return the layer the point lies in.
   * @param p The point.
   * @return The layer the Point is found in.
   */
  virtual unsigned int getLayer(Point p) const;
```

and the same is true for NearestPointBase

 ```
/**
   * Given a Point return the integral value associated with the layer
   * that point falls in for the layered average closest to that
   * point.
   *
   * @param p The point to look for in the layers.
   */
  virtual Real spatialValue(const Point & p) const override;

  /**
   * Get the points at which the nearest operation is performed
   * @return points
   */
  virtual const std::vector<Point> & getPoints() const { return _points; }

```

And InterfaceQpUserobjectBase
```

  /**
   * method returning the quadrature point value
   **/
  Real getQpValue(const dof_id_type elem, const unsigned int side, unsigned int qp) const;
  /**
   * function returning the element side average value
   **/
  Real getSideAverageValue(const dof_id_type elem, const unsigned int side) const;
```

and really just about any user object I see.
Am I misunderstanding how I should retrieve the value of the Element Integral Variable user object?

Thanks in advance

**Comment by GiudGiud** [comment, marked as answer]
Hello

we have an ongoing maintenance task to make this routine const. 

It was not because it used to be performing réduction across processes but that got cleared recently 

For now you ll want to const cast away the constness to be able to use getValue

Guillaume

**Reply by gabburgio** [reply]
Thank you!
